### ISSUE TITLE
Karto strange memory usage behaviour

### ISSUE THREAD

**Issue by zoidzilla** [SteveMacenski/slam_toolbox#558]
Hello,

I am running Karto and I am monitoring the RES memory using the [cpu_monitor](https://github.com/alspitz/cpu_monitor) package. I have used Karto individually (no global optimization and loop closures) and as a SLAM package (Karto SLAM, Karto with Ceres, Karto with g2o, Kator with GTSAM). What I am getting is the following figure (as extracted in Karto with Ceres as a back end, exhibiting similar behavior for all the other implementations). Furthermore, regarding SLAM Toolbox, I found in the following publication/thesis that exhibits similar Memory behavior since it is based on open Karto (  in figures 5.16, 5.17, 5.18 in [Optimal Wheelchair Multi-LiDAR Placement for Indoor SLAM](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiB5-PBhub7AhVvMVkFHa9wCcMQFnoECAsQAQ&url=https%3A%2F%2Frepositorio-aberto.up.pt%2Fbitstream%2F10216%2F135766%2F2%2F489043.pdf&usg=AOvVaw3_QVHbqdgi46SZCt7AoBSV)).  Now, I understand that the memory usage builds up. What I don't understand is the memory behavior after 70MB (it continually allocates and frees a lot of memory blocks). So far, I have not obtained any good explanation why this happens. Can you help?

Best regards,

![Karto_SLAM_Mem1](https://user-images.githubusercontent.com/2331748/206153274-898b0de3-f70f-40a2-a23b-398c7f1df327.jpg)

**Comment by SteveMacenski** [comment]
What if you turn off loop closures and anything related to the optimizer? I suspect it’s related to those

**Comment by zoidzilla** [comment]
Hello, I have already tried that as I am mentioning in my initial post: "I have used Karto individually (no global optimization and loop closures) and as a SLAM package", but it seams that it is completely independent from what back-end library you are using. My understanding so far it that this is totally due to Karto implementation. Now, since SLAM Toolbox tries to deal with the lifelong SLAM problem (and thus the total resources used),  maybe you or any of the involved developers have encounter this behavior and have an explanation to it. In any case any help would be greatly appreciate it!

**Comment by SteveMacenski** [comment]
Change the occupancy grid generation frequency (or just dont generate one) - does that change the pattern proportionally? It could be that when we reallocate the occupancy grid to populate it after new data is processed is taking up a chunk of memory to store then its deleted when we publish it to the ether. 

Both the regularity of the oscillation and the fact that it seems to grow quickly at first but then level off seems like the behavior I might expect from a SLAM trial in a space / map size.

I've definitely noticed this pattern (see charts in the readme) but I didn't concern myself much with it since it was still well below my requirements.

**Comment by zoidzilla** [comment]
Hi, that is actually a very probable explanation!! I could put a delay in the map update function to see what happens. Let me do some tests tomorrow and I will get back with my findings. Thanks a lot!

**Comment by zoidzilla** [comment]
Hello,

I did some tests and it was exactly as you said! I have attached three figures below. In figure 1 I have completely deactivated the occupancy grid deletion at every map update (so we have a constant build up). In the second figure, I have introduced a 5 sec delay exactly after I delete the occupancy grid in the map update function (here I have not changed any other parameters for loop closure detection and thus the algorithm stopped detecting global loop closures due to the 5sec delayed distance the robot traveled). Figure 3 I have introduced a 5 sec delay before I delete the occupancy grid map. The only strange issue here is that theoretically I would have expected to have very small negative peaks that would reflect the occupancy grid deletion. It seams that there is some kind of overlap when creating a new occupancy grid map immediately after when deleting one. In any case, I now understand why this is happening and thank you very much for your assistance! All the best!
   
![Fig11_no_grid_delete](https://user-images.githubusercontent.com/2331748/206734119-49277a49-bb29-452d-ade6-37ea6ac0c19e.jpg)
Figure 1
![Fig11_5s_delay_after_delete](https://user-images.githubusercontent.com/2331748/206734200-8b8b1468-5942-4868-8a05-8b64488002f1.jpg)
Figure 2
![Fig11_5s_delay_before_delete](https://user-images.githubusercontent.com/2331748/206734288-be19616e-9347-4b33-a810-a4b4fb2f65ec.jpg)
Figure 3
